diff-tests/core/core_test.go

```diff
@@ -0,0 +1,458 @@
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Staking module does contain redelegations
+	s.providerStakingKeeper().IterateRedelegations(s.ctx(P),
+		func(index int64, ubd stakingtypes.Redelegation) bool {
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Staking module does not contain unbonding validators
+	endTime := time.Unix(math.MaxInt64, 0)
+	endHeight := int64(math.MaxInt64)
+	unbondingValIterator := s.providerStakingKeeper().ValidatorQueueIterator(s.ctx(P), endTime, endHeight)
+	defer unbondingValIterator.Close()
+	for ; unbondingValIterator.Valid(); unbondingValIterator.Next() {
+		s.T().Fatal(FAIL_MSG)
+	}
+
+	// Consumer has no slash requests
+	s.Require().Empty(s.consumerKeeper().GetPendingSlashRequests(s.ctx(C)))
+
+	// Consumer has no maturities
+	s.consumerKeeper().IteratePacketMaturityTime(s.ctx(C),
+		func(vscId uint64, timeNs uint64) bool {
+			s.T().Fatal(FAIL_MSG)
+			return false // Don't stop
+		})
+
+	// Consumer power
+	for i := 0; i < len(initState.ValStates.Status); i++ {
+		expectFound := initState.ValStates.Status[i] == stakingtypes.Bonded
+		expectPower := initState.ValStates.Tokens[i]
+		addr := s.validator(int64(i))
+		val, found := s.consumerKeeper().GetCCValidator(s.ctx(C), addr)
+		s.Require().Equal(expectFound, found)
+		if expectFound {
+			if int64(expectPower) != val.Power {
+				s.T().Fatal(FAIL_MSG)
+			}
+		}
+	}
+
+	// The offset time is the last committed time, but the SUT is +1 block ahead
+	// because the currentHeader time is ahead of the last committed. Therefore sub
+	// the difference (duration of 1 block).
+	s.Require().Equal(int64(s.offsetTimeUnix), s.time(P).Add(time.Second*(-6)).Unix())
+	s.Require().Equal(int64(s.offsetTimeUnix), s.time(C).Add(time.Second*(-6)).Unix())
+
+	// The offset height is the last committed height, but the SUT is +1 because
+	// the currentHeader is +1 ahead of the last committed. Therefore sub 1.
+	s.Require().Equal(s.offsetHeight, s.height(P)-1)
+	s.Require().Equal(s.offsetHeight, s.height(C)-1)
+
+	// Network is empty
+	s.Require().Empty(s.simibc.Link.OutboxPackets[P])
+	s.Require().Empty(s.simibc.Link.OutboxPackets[C])
+	s.Require().Empty(s.simibc.Link.OutboxAcks[P])
+	s.Require().Empty(s.simibc.Link.OutboxAcks[C])
+}
+
+// Test a set of traces
+func (s *CoreSuite) TestTraces() {
+	s.traces = Traces{
+		Data: LoadTraces("traces.json"),
+	}
+	for i := range s.traces.Data {
+		s.Run(fmt.Sprintf("Trace num: %d", i), func() {
+			// Setup a new pair of chains for each trace
+			s.SetupTest()
+			s.traces.CurrentTraceIx = i
+			defer func() {
+				// If a panic occurs, we trap it to print a diagnostic
+				// and improve debugging experience.
+				if r := recover(); r != nil {
+					fmt.Println(s.traces.Diagnostic())
+					fmt.Println(r)
+					panic("Panic occurred during TestTraces")
+				}
+			}()
+			// Record information about the trace, for debugging
+			// diagnostics.
+			s.executeTrace()
```

Comment:
Where does the state after the action get checked (the "Consequence")? I see `matchState` being called in the `endAndBeginBlock`. Is that it? If it is, it seems odd to have your state being checked by one of your actions.

Reply:
We only compare committed block state because that's the only thing that matters at the end of the day. Anything that happens inside the block is implementation detail. I.e. as long as the same transactions result in the same committed block state we don't care what the go implementation actually does.

Reply:
We could check after every action but IMO that would bring unnecessary restrictions and make it harder to make the model abstract and to make the implementation efficient